pennylane/interfaces/batch/__init__.py

```diff
@@ -16,14 +16,59 @@
 capabilities with different machine learning libraries.
 """
 # pylint: disable=import-outside-toplevel,too-many-arguments,too-many-branches
+import contextlib
 from functools import wraps
+import itertools
 
 from cachetools import LRUCache
 import numpy as np
 
 import pennylane as qml
 
 
+INTERFACE_NAMES = {
+    "NumPy": (None,),
+    "Autograd": ("autograd", "numpy"),  # for backwards compatibility
+    "PyTorch": ("torch", "pytorch"),
+    "TensorFlow": ("tf", "tensorflow"),
+}
+"""dict[str, str]: maps allowed interface strings to the name of the interface"""
+
+SUPPORTED_INTERFACES = list(itertools.chain(*INTERFACE_NAMES.values()))
+
+
+@contextlib.contextmanager
+def set_shots(device, shots):
+    """Context manager to temporarily change the shots
+    of a device.
+
+    This context manager can be used in two ways.
+
+    As a standard context manager:
+
+    >>> dev = qml.device("default.qubit", wires=2, shots=None)
+    >>> with set_shots(dev, shots=100):
+    ...     print(dev.shots)
+    100
+    >>> print(dev.shots)
+    None
+
+    Or as a decorator that acts on a function that uses the device:
+
+    >>> set_shots(dev, shots=100)(lambda: dev.shots)()

```

Comment:
I'm a bit confused by this example. Is the idea that you can do something like this?

```python
@qml.set_shots(100)
@qml.qnode(dev)
def circuitI():
    ....

```

Reply:
Oh never mind I see later how it is used to act on the device's `batch_execute` function. What confused me is the `lambda` here.

Reply:
> Is the idea that you can do something like this?

Good question 🤔 That probably works, but it was not my intention for that UX to work lol. This is just a small utility function for the batch execution pipeline, rather than something I meant to be user facing.

In terms of the code snippet in the example --- this confused me as well to be honest. The [`contextlib`](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator) docs say that all context managers can also be used as decorators, but they don't provide an example.

Note that the following _also_ works,

```python
set_shots(dev, shots=100)(lambda dev: dev.shots)(dev)
```

so there seem to be a number of permutations that work.

Reply:
Double checked, it doesn't work 😆 

![image](https://user-images.githubusercontent.com/2959003/132279546-ba122efe-4880-458a-835c-aa8a44471e54.png)

Reply:
Ohhhh it is because, by default, the QNode is 'swapping' the device under the hood to use backprop. If you force parameter-shift, it... works 🤯 

![image](https://user-images.githubusercontent.com/2959003/132280215-9c7c50fb-834c-4203-bff6-50f265117194.png)

Reply:
@glassnotes this is a cool unintentional feature!